### PR TITLE
Revamp `sf::Utf<N>` tests

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

Related to https://github.com/SFML/SFML/issues/3406

In the same spirit as #3418 I basically rewrote all the `sf::Utf` unit tests originally added in #3328. When I first wrote these tests I did not understand Unicode in general or our specific implementation of `sf::Utf`. I now have much better understanding of what all these functions do and feel confident these tests do a great job capturing the current functionality. 

Along the way I discovered a few things that needed to change. clang-tidy flagged some missing `const`s so I had to add those and I also removed an unnecessary function declaration from Utf.hpp. These two changes are in their own atomic commits so you can clearly see those changes separate from the new tests.

On top of that I elected to add `static_assert`s which check that we're using these functions correctly. For example `sf::Utf8::fromWide` needs to be called with a wide string so I assert that the provided iterators point to somethign the same size as a `wchar_t`. In the future I may make the asserts more strict and assert for specifically `wchar_t` but this is a decent starting point that checks for the most egregious misuses. Turns out the original tests I wrote used the wrong input data and these asserts caught that :) Eventually I would like to add similar `static_assert`s for output iterators types but that's a bit more complicated and will have to wait.

In writing these tests I discovered a bug in `sf::priv::copy` (renamed to `sf::priv::copyBits` to better reflect its new functionality) and wrote a huge comment block explaining why I had to make some unintuitive changes to fix that so I hope that's clear. This bug went unnoticed because until now we were not testing anything with Latin1 strings.

As for shortcomings, I these tests don't really cover error case behavior. Those tests probably ought to be added but I thought it was out of scope for this PR. I also skipped doing anything with locales. Certain functions have default `std::locale` parameters. I never provided non-default values so that's something to potentially do in the future. The last major shortcoming is due to MinGW. It produces different result when converting characters than all other compilers which limits what strings we can test. This is why there are no Latin1 characters in our tests for example. ASCII and Emoji work fine but things like `é` or Polish `ń` (outside of Latin1) don't get converted the same way as other compilers. I don't know why this is the case. Perhaps we can fix this with careful locale manipulation but I'm really not sure. MinGW just continues to be a thorn in my side.

---

These tests provide more great groundwork for future Unicode work. These tests don't go too far towards testing error cases but I wager there are some edge case bugs to be found as we dig deeper. I also suspect now that we have these tests I may want to clean up the implementation of `sf::Utf` a bit.

### CODE REVIEW

File: include/SFML/System/Utf.inl
Comment (ZXShady) [inline include/SFML/System/Utf.inl:1]: Missing include for `<iterator>` it uses `std::iterator_traits`

**Review by ZXShady — CHANGES_REQUESTED** [review]

**Review by ZXShady — APPROVED** [review]

**Review by eXpl0it3r — APPROVED** [review]
I trust your new gained knowledge and the test suite for this bit moving to be correct. 😄

File: test/System/Utf.test.cpp [hunk 1]
```diff
@@ -4,237 +4,608 @@
...
 #include <string_view>
 
+namespace
+{
+// Return either argument depending on whether wchar_t is 16 or 32 bits
+// Lets us write tests that work on both Windows where wchar_t is 16 bits
+// and elsewhere where it is 32. Otherwise the tests would only work on
+// one OS or the other.
+template <typename T>
+auto select(const std::basic_string_view<T>& string16, const std::basic_string_view<T>& string32)
```
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/Utf.test.cpp:14]: Good idea. I like this name.
